lib/ansible/plugins/callback/grafana_annotations.py

```diff
@@ -0,0 +1,261 @@
+        description: Grafana password used for authentication. Ignored if grafana_api_key is provided.
+        env:
+          - name: GRAFANA_PASSWORD
+        ini:
+          - section: callback_grafana_annotations
+            key: grafana_password
+        default: ansible
+      grafana_dashboard_id:
+        description: The grafana dashboard id where the annotation shall be created.
+        env:
+          - name: GRAFANA_DASHBOARD_ID
+        ini:
+          - section: callback_grafana_annotations
+            key: grafana_dashboard_id
+      grafana_panel_id:
+        description: The grafana panel id where the annotation shall be created.
+        env:
+          - name: GRAFANA_PANEL_ID
+        ini:
+          - section: callback_grafana_annotations
+            key: grafana_panel_id
+"""
+
+
+PLAYBOOK_START_TXT = """\
+Started playbook {playbook}
+
+From '{hostname}'
+By user '{username}'
+"""
+
+PLAYBOOK_ERROR_TXT = """\
+Playbook {playbook} Failure !
+
+From '{hostname}'
+By user '{username}'
+
+'{task}' failed on {host}
+
+debug: {result}
+"""
+
+PLAYBOOK_STATS_TXT = """\
+Playbook {playbook}
+Duration: {duration}
+Status: {status}
+
+From '{hostname}'
+By user '{username}'
+
+Result:
+{summary}
+"""
+
+
+def to_millis(dt):
+    return int(dt.strftime('%s')) * 1000
+
+
+class CallbackModule(CallbackBase):
+    """
+    ansible grafana callback plugin
+    ansible.cfg:
+        callback_plugins   = <path_to_callback_plugins_folder>
+        callback_whitelist = grafana_annotations
+    and put the plugin in <path_to_callback_plugins_folder>
+    """
+
+    CALLBACK_VERSION = 1.0
+    CALLBACK_TYPE = 'aggregate'
+    CALLBACK_NAME = 'grafana_annotations'
+    CALLBACK_NEEDS_WHITELIST = True
+
+    def __init__(self, display=None):
+
+        super(CallbackModule, self).__init__(display=display)
+
+        self.headers = {'Content-Type': 'application/json'}
+        self.force_basic_auth = False
+        self.hostname = socket.gethostname()
+        self.username = getpass.getuser()
+        self.start_time = datetime.now()
+        self.errors = 0
+
+    def set_options(self, task_keys=None, var_options=None, direct=None):
+
+        super(CallbackModule, self).set_options(task_keys=task_keys, var_options=var_options, direct=direct)
+
+        self.grafana_api_key = self.get_option('grafana_api_key')
+        self.grafana_url = self.get_option('grafana_url')
+        self.validate_grafana_certs = self.get_option('validate_grafana_certs')
+        self.http_agent = self.get_option('http_agent')
+        self.grafana_user = self.get_option('grafana_user')
+        self.grafana_password = self.get_option('grafana_password')
+        self.dashboard_id = self.get_option('grafana_dashboard_id')
+        self.panel_id = self.get_option('grafana_panel_id')
+
+        if self.grafana_api_key:
+            self.headers['Authorization'] = "Bearer %s" % self.grafana_api_key
+        else:
+            self.force_basic_auth = True
```

Comment:
Is it possible to add a check for `grafna_user` and `grafana_password` options here ? with a warning message "You must provide a grafana_api_key or a grafana_user + grafana_password option".
The rest Looks good to me

Reply:
There are defaults to grafana_user and grafana_password

Reply:
My bad, I've missed it !